src/lib/ensemble/bagging.ts

```diff
@@ -0,0 +1,207 @@
+/**
+ * References:
+ * - https://scikit-learn.org/stable/modules/generated/sklearn.ensemble.BaggingClassifier.html
+ */
+import { DecisionTreeClassifier } from '../tree';
+import { Type1DMatrix, Type2DMatrix } from '../types';
+import math from '../utils/MathExtra';
+import { ensure2DMatrix, inferShape } from '../utils/tensors';
+import { validateFitInputs } from '../utils/validation';
+
+/**
+ * A Bagging classifier is an ensemble meta-estimator that fits
+ * base classifiers each on random subsets of the original dataset
+ * and then aggregate their individual predictions by voting
+ * to form a final prediction
+ * @example
+ * const classifier = new BaggingClassifier({
+ *  baseEstimator: LogisticRegression,
+ *  maxSamples: 1.0,
+ * });
+ * const X = [[1], [2], [3], [4], [5]];
+ * const y = [1, 1, 1, 1, 1];
+ * classifier.fit(X, y);
+ * classifier.predict(X);
+ */
+export class BaggingClassifier {
+  private BaseEstimator: any;
+  private numEstimators: number;
+  private estimatorOptions: any;
+  private maxSamples: number;
+  private maxFeatures: number;
+  private bootstrapSamples: boolean;
+  private bootstrapFeatures: boolean;
+  private estimators: any[] = [];
+  private estimatorsFeatures: number[][] = [];
+  private maxSamplesIsFloating: boolean = true;
+  private maxFeaturesIsFloating: boolean = true;
+
+  /**
+   * @param BaseEstimator - The model that will be used as a basis of ensemble.
+   * @param numEstimators - The number of estimators that will be used in ensemble.
+   * @param maxSamples - The number of samples to draw from X to train each base estimator.
+   *  Is used in conjunction with @param maxSamplesIsFloating.
+   *  If @param maxSamplesIsFloating is false, then draw maxSamples samples.
+   *  If @param maxSamplesIsFloating is true, then draw max_samples * shape(X)[0] samples.
+   * @param maxFeatures - The number of features to draw from X to train each base estimator.
+   *  Is used in conjunction with @param maxFeaturesIsFloating
+   *  If @param maxFeaturesIsFloating is false, then draw max_features features.
+   *  If @param maxFeaturesIsFloating is true, then draw max_features * shape(X)[1] features.
+   * @param bootstrapSamples - Whether samples are drawn with replacement. If false, sampling without replacement is performed.
+   * @param bootstrapFeatures - Whether features are drawn with replacement.
+   * @param estimatorOptions - constructor options for BaseEstimator.
```

Comment:
Missing docs for `maxSamplesIsFloating` and `maxFeaturesIsFloating`

Reply:
Their meaning is described in maxSamples and maxFeatures

Reply:
Ok, let me pull them out so it can be picked up by the doc generator

Reply:
Ok resolved